### PR TITLE
chore(headless): dynamically load reducers

### PR DESCRIPTION
**Pull request by samisayegh**
The goal of this PR is to simplify Headless configuration. Rather than asking a developer to specify reducers, they would be added dynamically when instantiating a controller. Some benefits of the approach are:

- Less configuration for Typescript users. The `reducers` option when instantiating an engine would eventually become optional.
- Reduces the chance of errors for javascript users. The state of the engine argument on controllers can be empty.
- Smaller bundles. Only the instantiated controllers and their corresponding reducers will be included in the bundle. At the moment, all the reducers in `searchAppReducers` are included, whether or not the feature set is being used.

Inspiration: https://redux.js.org/recipes/code-splitting#using-a-reducer-manager

### CODE REVIEW

File: packages/headless/src/app/headless-engine.ts [hunk 1]
```diff
@@ -281,6 +291,11 @@ export class HeadlessEngine<Reducers extends ReducersMapObject>
     }
   }
 
+  public addReducers(reducers: ReducersMapObject) {
+    this.reducerManager.add(reducers);
+    this.reduxStore.replaceReducer(this.reducerManager.reducer);
```
Comment (olamothe) [inline packages/headless/src/app/headless-engine.ts:296]: Small nitpick, but shouldn't the reducer manager be in charge of calling `replaceReducer` internally on the redux store, when `add` is executed ?

Seems natural to me that a manager would also manage that part.

File: packages/headless/src/controllers/facets/facet/headless-facet.ts [hunk 2]
```diff
@@ -330,7 +338,7 @@ export function buildFacet(
       const request = getRequest();
       const response = getResponse();
 
-      const isLoading = engine.state.search.isLoading;
+      const isLoading = getIsLoading();
```
Comment (olamothe) [inline packages/headless/src/controllers/facets/facet/headless-facet.ts:341]: Is this change needed or just an improvement you saw as you were working on this ?

If it's needed, I'd like to understand why ;)

**Review by olamothe — APPROVED** [review]
Very clean !

File: packages/headless/src/controllers/facets/facet/headless-facet.ts (hunk 2 above)
Comment (samisayegh) [reply packages/headless/src/controllers/facets/facet/headless-facet.ts:341]: Yep, I made this change to resolve a typescript error. Inside getters only, typescript doesn't correctly infer the state of the engine. It sees it as `unknown`.

Everywhere else in the controller, ts understands the correct type via the typeguard.

```
buildSomeController(
engine: Engine<unknown>
) {
    // engine state is "unknown"
    if (!loadReducers(engine)) {
        throw Error
    }

    // if we are here, typescript understands that the engine state is that specified by the loadReducers type guard

    return {
        get state() {
           // here, typescript goes back to thinking the engine state is "unknown" for some reason.
       }
    }
}
```

File: packages/headless/src/app/headless-engine.ts [hunk 3]
```diff
@@ -238,11 +239,18 @@ export interface Engine<State = SearchAppState> {
...
+   */
   renewAccessToken: () => Promise<string>;
   /**
    * The logger instance used by headless.
    * */
   logger: Logger;
+  /**
+   * Adds the specified reducers to the store.
+   */
+  addReducers(reducerMap: ReducersMapObject): void;
```
Comment (nathanlb) [inline packages/headless/src/app/headless-engine.ts:253]: The argument is called `reducerMap` here and `reducers` on line 294. For consistency maybe they should be the same.

**Review by nathanlb — COMMENTED** [review]
This looks good but I'm still a bit confused at what this translates to on the ui library side. When instantiating the engine no reducers are required now I guess. I'm probably just not searching right but I don't see where they're loaded dynamically based on registered components.

File: packages/headless/src/app/headless-engine.ts (hunk 1 above)
Comment (samisayegh) [reply packages/headless/src/app/headless-engine.ts:296]: Hmm, I see ReducerManager's role as taking care of holding, adding and building the final `reducer` function.

I feel giving it the role of replacing the reducer on the store would be a tad too much. The `reduxStore` instance is part of this class. I could pass it into ReducerManager, but I feel it would be clearer to not do that,

File: packages/headless/src/app/reducer-manager.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  add: (newReducers: ReducersMapObject) => void;
+}
+
+export function createReducerManager(
+  initialReducers: ReducersMapObject
+): ReducerManager {
+  const reducers = {...initialReducers};
+
+  return {
+    get reducer() {
```
Comment (ThibodeauJF) [inline packages/headless/src/app/reducer-manager.ts:14]: I would be more explicit to name this `combinedReducer`

**Review by ThibodeauJF — APPROVED** [review]

**Comment by samisayegh** [comment]
> This looks good but I'm still a bit confused at what this translates to on the ui library side. When instantiating the engine no reducers are required now I guess. I'm probably just not searching right but I don't see where they're loaded dynamically based on registered components.


Yea, once we convert all the controllers to the new approach, reducers would not need to be specified. The dynamic load happens at the [start of the controller](https://github.com/coveo/ui-kit/pull/709/files#diff-19560fb69917a92e382a82c7948d49196ea8dafbc269b614feb556655608c8b4R144), where it adds the necessary reducers to the passed engine.

**Comment by nathanlb** [comment]
Oh I see so this is added to the controllers. Sounds good 👍  thanks!

**Review by nathanlb — APPROVED** [review]

File: packages/headless/src/controllers/facets/facet/headless-facet.ts (hunk 2 above)
Comment (samisayegh) [reply packages/headless/src/controllers/facets/facet/headless-facet.ts:341]: An alternative would be to do this, and update references to use `resolvedEngine` inside the controller:

```
export function buildFacet(engine: Engine<unknown>, props: FacetProps): Facet {
  if (!loadFacetReducers(engine)) {
    throw loadReducerError;
  }

  const resolvedEngine = engine;
```

Down-side is one would be able to use `engine` or `resolvedEngine`, except inside getters where only `resolvedEngine` would work.

Comment (samisayegh) [reply packages/headless/src/controllers/facets/facet/headless-facet.ts:341]: I created a report for the TS team: https://github.com/microsoft/TypeScript/issues/43676
